RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc

```diff
@@ -0,0 +1,632 @@
+  nMaxItersNNLS_ = iNMaxItersNNLS;
+
+  deltaChiSqThresh_ = iDeltaChiSqThresh;
+  nnlsThresh_    = iNnlsThresh;
+
+  bxOffsetConf_ = -(*std::min_element(activeBXs_.begin(), activeBXs_.end()));
+  bxSizeConf_   = activeBXs_.size();
+
+}
+
+void MahiFit::phase1Apply(const HBHEChannelInfo& channelData,
+			  float& reconstructedEnergy,
+			  float& reconstructedTime,
+			  bool& useTriple, 
+			  float& chi2) const {
+
+  assert(channelData.nSamples()==8||channelData.nSamples()==10);
+
+  resetWorkspace();
+
+  nnlsWork_.tsSize = channelData.nSamples();
+  nnlsWork_.tsOffset = channelData.soi();
+  nnlsWork_.fullTSOffset = fullTSofInterest_ - nnlsWork_.tsOffset;
+
+  // 1 sigma time constraint
+  if (channelData.hasTimeInfo()) nnlsWork_.dt=timeSigmaSiPM_;
+  else nnlsWork_.dt=timeSigmaHPD_;
+
+  //Dark current value for this channel (SiPM only)
+  float darkCurrent =  getSiPMDarkCurrent(channelData.darkCurrent(), 
+					  channelData.fcByPE(),
+					  channelData.lambda());
+
+  //Average pedestal width (for covariance matrix constraint)
+  float pedVal = 0.25*( channelData.tsPedestalWidth(0)*channelData.tsPedestalWidth(0)+
+			channelData.tsPedestalWidth(1)*channelData.tsPedestalWidth(1)+
+			channelData.tsPedestalWidth(2)*channelData.tsPedestalWidth(2)+
+			channelData.tsPedestalWidth(3)*channelData.tsPedestalWidth(3) );
+
+  nnlsWork_.pedConstraint = pedVal*SampleMatrix::Ones(nnlsWork_.tsSize, nnlsWork_.tsSize);
+  nnlsWork_.amplitudes.resize(nnlsWork_.tsSize);
+  nnlsWork_.noiseTerms.resize(nnlsWork_.tsSize);
+
+  std::vector<float> reconstructedVals;
+  
+  double tsTOT = 0, tstrig = 0; // in GeV
+  for(unsigned int iTS=0; iTS<nnlsWork_.tsSize; ++iTS){
+    double charge = channelData.tsRawCharge(iTS);
+    double ped = channelData.tsPedestal(iTS);
+
+    nnlsWork_.amplitudes.coeffRef(iTS) = charge - ped;
+
+    //ADC granularity
+    double noiseADC = (1./sqrt(12))*channelData.tsDFcPerADC(iTS);
+
+    //Dark current (for SiPMs)
+    double noiseDC=0;
+    if(channelData.hasTimeInfo() && !channelData.hasEffectivePedestals() && (charge-ped)>channelData.tsPedestalWidth(iTS)) {
+      noiseDC = darkCurrent;
+    }
+
+    //Photostatistics
+    double noisePhoto = 0;
+    if ( (charge-ped)>channelData.tsPedestalWidth(iTS)) {
+      noisePhoto = sqrt((charge-ped)*channelData.fcByPE());
+    }
+
+    //Electronic pedestal
+    double pedWidth = channelData.tsPedestalWidth(iTS);
+
+    //Total uncertainty from all sources
+    nnlsWork_.noiseTerms.coeffRef(iTS) = noiseADC*noiseADC + noiseDC*noiseDC + noisePhoto*noisePhoto + pedWidth*pedWidth;
+
+    tsTOT += charge - ped;
+    if( iTS==nnlsWork_.tsOffset ){
+      tstrig += (charge - ped)*channelData.tsGain(0);
+    }
+  }
+
+  if(tstrig >= ts4Thresh_ && tsTOT > 0) {
+
+    useTriple=false;
+
+    // only do pre-fit with 1 pulse if chiSq threshold is positive
+    if (chiSqSwitch_>0) {
+      doFit(reconstructedVals,1);
+      if (reconstructedVals[1]>chiSqSwitch_) {
```

Comment:
reconstructedVals is not promised to have size()>=2 in doFit.
a much clearer interface for do fit would be 

bool doFit( std::array<float,3> & reconstructedVals, int ibx)

such than the success or failure comes from a simple bool check and size of the output variable is known by all parties.

Reply:
We don't really have a concept of "success" or "failure" fits, so I didn't add back the boolean return, but I did change to the array<float,3>.

Reply:
is the index correct here?
Given the last bugfix it is clear that reconstructedVals[1] contains time and not chi2.
Incidentally, ``` chiSqSwitch       = cms.double(-10.0) ``` and will most likely have no effect
.. except that I do not quite understand which values we get if ```foundintime``` in ```doFit``` is false and the default (uninitialized) values of reconstructedVals show up here where a check on reconstructedVals is made: the result can be random

Reply:
Yes thanks for catching that! 
And the chiSqSwitch is not currently used, but it's meant to preserve the functionality in method 2 to switch between 1 and 3 pulse fits (in Mahi it switches between 1 pulse and the python configured bxvector), and allow for pre-fitting (like Ecal Multifit).